module/zfs/zio.c

```diff
@@ -3526,7 +3526,25 @@ zio_vdev_io_start(zio_t *zio)
 			    ZIO_FLAG_RESILVER | ZIO_FLAG_INDUCE_DAMAGE));
 		}
 	}
-
+	/*
+	 * We keep track of time-sensitive I/Os so that the scan thread
+	 * can quickly react to certain workloads.  In particular, we care
+	 * about non-scrubbing, top-level reads and writes with the following
+	 * characteristics:
+	 *      - synchronous writes of user data to non-slog devices
+	 *      - any reads of user data
+	 * When these conditions are met, adjust the timestamp of spa_last_io
+	 * which allows the scan thread to adjust its workload accordingly.
+	 */
+	if (!(zio->io_flags & ZIO_FLAG_SCAN_THREAD) && zio->io_bp != NULL &&
+	    vd == vd->vdev_top && !vd->vdev_islog &&
+	    zio->io_bookmark.zb_objset != DMU_META_OBJSET &&
+	    zio->io_txg != spa_syncing_txg(spa)) {
+		uint64_t old = spa->spa_last_io;
+		uint64_t new = NSEC2MSEC(gethrtime());
+		if (old != new)
+			(void) atomic_cas_64(&spa->spa_last_io, old, new);
+	}
```

Comment:
This logic all predates the addition of the new write throttle and vdev queuing code.  Rather that simply reinstating the previous logic, I'm sure we can do something a little better now.  That would be desirable since if I recall correctly this throttle mechanism was difficult to tune in practice and frequently resulted in sub-optimal scrub and resilver times since it could starve out scrub/resilver I/Os.

One possibly worth exploring would be to update `vdev_queue_class_to_issue()` to not issue pending scrub I/O when there are outstanding or pending synchronous writes or reads.  I'd expect that moving the throttle itself closer to the device would improve our overall disk utilization.  Furthermore, you might not want to throttle for non-rotation devices and this is something we could take in to account at this lower layer.

@tcaputi and @amotin may have some additional thoughts about this.

Reply:
I'm interested to see some performance numbers / pathological cases here. Back when I was working on this code, we tested it fairly extensively (since the whole point was to make resilvers and scrubs more predictable and faster when there was headroom for it). I agree that the old mechanism had a good number of issues, which was why we removed it when we rewrote the scanning code.

I'm particularly interested in the example @snajpa was talking about. Holding IO's in a queue for 34 seconds definitely seems like a bug.

Reply:
@behlendorf My only concern at the moment is working through the diff between FreeBSD and master. Should I close this PR and revert the change in ZoF, trusting that any underlying issues will be dealt with elsewhere?

Reply:
I think before we make any change in this area we need to understand and characterize the performance implications of that change.  I don't mind leaving this open as a WIP if someone wants to use this as a starting point for that investigation.